packages/react-dom-bindings/src/server/ReactDOMServerExternalRuntime.js

```diff
@@ -21,35 +21,37 @@ if (!window.$RC) {
   window.$RM = new Map();
 }
 
-if (document.readyState === 'loading') {
-  if (document.body != null) {
+if (document.body != null) {
+  if (document.readyState === 'loading') {
     installFizzInstrObserver(document.body);
-  } else {
-    // body may not exist yet if the fizz runtime is sent in <head>
-    // (e.g. as a preinit resource)
-    // $FlowFixMe[recursive-definition]
-    const domBodyObserver = new MutationObserver(() => {
-      // We expect the body node to be stable once parsed / created
-      if (document.body) {
-        if (document.readyState === 'loading') {
-          installFizzInstrObserver(document.body);
-        }
-        handleExistingNodes();
-        // We can call disconnect without takeRecord here,
-        // since we only expect a single document.body
-        domBodyObserver.disconnect();
-      }
-    });
-    // documentElement must already exist at this point
-    // $FlowFixMe[incompatible-call]
-    domBodyObserver.observe(document.documentElement, {childList: true});
   }
-}
+  // $FlowFixMe[incompatible-cast]
+  handleExistingNodes((document.body /*: HTMLElement */));
+} else {
+  // Document must be loading -- body may not exist yet if the fizz external
+  // runtime is sent in <head> (e.g. as a preinit resource)
+  // $FlowFixMe[recursive-definition]
+  const domBodyObserver = new MutationObserver(() => {
+    // We expect the body node to be stable once parsed / created
+    if (document.body != null) {
+      if (document.readyState === 'loading') {
+        installFizzInstrObserver(document.body);
+      }
+      // $FlowFixMe[incompatible-cast]
+      handleExistingNodes((document.body /*: HTMLElement */));
 
-handleExistingNodes();
+      // We can call disconnect without takeRecord here,
+      // since we only expect a single document.body
+      domBodyObserver.disconnect();
+    }
+  });
+  // documentElement must already exist at this point
+  // $FlowFixMe[incompatible-call]
+  domBodyObserver.observe(document.documentElement, {childList: true});
+}
 
-function handleExistingNodes() {
-  const existingNodes = document.getElementsByTagName('template');
+function handleExistingNodes(target /*: HTMLElement */) {
+  const existingNodes = Array.from(target.children);
```

Comment:
I'm pretty sure `querySelectorAll` will outperform this by a lot. In practice it may not make a huge difference given the (un)popularity of template in many sites

Reply:
Maybe not by a lot but I did find a jsperf type benchmark that showed it was faster

Reply:
Thanks for the feedback here! I had incorrectly assumed that querySelectorAll might be slower for a deep + widely branching tree, and that `document.body.children` was likely small. Since I'm not at all familiar with browser DOM perf, will go ahead and change this.

I ran some ad-hoc timing on my laptop (on `DOMContentLoaded` event for facebook home page, a 'representative DOM sample'), and it looks like `querySelectorAll('template')` is generally faster.

Safari:
<img width="500" alt="Pasted image 20230304110032" src="https://user-images.githubusercontent.com/34200447/222916287-861886e5-61ca-426d-b8d9-cf127b47e685.png">

Chrome:
<img width="500" alt="image" src="https://user-images.githubusercontent.com/34200447/222916486-4f862f25-ad45-47be-8139-d4115c9308d0.png">